WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -239,16 +232,35 @@ def edit
     render :edit
   end
 
+  def clone_competition
+    @competition = Competition.new
+    cloned_competition = Competition.find(params[:id])
+    if cloned_competition
+      attributes = cloned_competition.attributes
+      # Don't clone the following attributes.
+      %w(id name cellName year month day endMonth endDay registration_open registration_close results_posted_at).each { |attribute| attributes.delete attribute }
+      @competition.assign_attributes(attributes)
+      @competition.organizers |= cloned_competition.organizers
+      @competition.delegates |= cloned_competition.delegates
+    end
+    if current_user.any_kind_of_delegate?
+      @competition.delegates |= [current_user]
+    end
+    @competition.showAtAll = false
+    @competition.isConfirmed = false
+    render :new
+  end
```

Comment:
I'd like to have this logic inside the model to slim down the controller as much as we can (it's big enough). In the current situation, instead of using the `before_validation` callback as before, it would be better just to make a method. Something like:
```ruby
# Called for the competition which we want to clone data from
def get_clone
  # All the logic here
  # Return a *new* competition with the attributes assigned
end
```

Then in the controller something like this:
```ruby
cloned_competition = Competition.find(params[:id])
@competition = cloned_competition.get_clone
```